Gh-398: Where to run gaffer administration guide (#406)
gchq/gaffer-doc: docs/administration-guide/gaffer-deployment/gaffer-docker/gaffer-images.md

baked into the image. This works well if the configuration itself is rather
static and the same across all environments.

Creating a custom image can also be useful if you want to load custom extensions
to use with Gaffer (e.g. Jars) by default.

To create a custom image simply make a new `Dockerfile` and use one of the Gaffer
images as the base image like the following:

```dockerfile
FROM gchq/gaffer-rest:latest

# Copy over the existing directory with store configs in
COPY ./custom/configs/gaffer/store /gaffer/store
```

Then build the new image using a suitable tool or just plain Docker from the
current directory like:

```bash
docker build -t my-gaffer-rest .
```

### Adding Additional Libraries

By default with the Gaffer deployment you get access to the:

- Sketches library
- Time library
- Bitmap Library
- JCS cache library

If you want more libraries than this (either one of ours of one of your own) you
will need to customise the docker images and use them in place of the defaults.

At the moment, the `gchq/gaffer-rest` image uses a runnable jar file located at
`/gaffer/jars`. When it runs it includes the `/gaffer/jars/lib` on the

> classpath. There is nothing in there by default because all the dependencies are

Reviewer: ```suggestion
classpath. This is empty by default because all the dependencies are
```

bundled in to the JAR. However, if you wanted to add your own jars, you can add
then to this directory like the following:

```dockerfile
FROM gchq/gaffer-rest:latest
COPY ./custom-lib:1.0-SNAPSHOT.jar /gaffer/jars/lib/
```

To add any libraries to the `gchq/gaffer` image in order to push down any extra

> value objects and filters to Accumulo you have to add the jars to the

Reviewer: I'm not sure what's meant by "push down any extra value objects and filters". This directory is used for JARs to add to the Accumulo classpath. Either new iterators or dependencies for them. There's more detail about this on the Accumulo migration page as this path differs between Accumulo 1 and 2. If this was copied from elsewhere it needs to be updated.

Author: This paragraph I believe was in the existing k8s docs that I moved here, I agree the wording is a little strange I'll simplify and add a link to the migration page

# Gaffer Images

> As demonstrated in the [quickstart](../quickstart.md) its very simple to start

Reviewer 2: ```suggestion
As demonstrated in the [quickstart](../quickstart.md) it is very simple to start
```

up a basic in memory gaffer graph using the available Open Container Initiative
(OCI) images.

For large scale graphs with persistent storage you will want to use a different
storage backend to a Map Store; the recommended one being Accumulo. To do this a
different deployment of containers are required. This guide will run through the
containers needed for a basic Accumulo cluster and how to configure and create
custom images of Gaffer.

## Available Images

Currently there are a few different images that can be used to run a Gaffer
deployment. The main ones are outlined in the following table and are all
available on [Docker Hub](https://hub.docker.com/u/gchq).

| Image | Description |
| ----- | ----------- |

> | `gchq/accumulo` | This image is a containerised deployment of [Apache Accumulo](https://accumulo.apache.org/). This was created as historically there has not been an available official image from the maintainers of Accumulo; however, there has since been an [offical image](https://github.com/apache/accumulo-docker) made available but it is not currently in use in Gaffer. |

Reviewer 3: I think the official image is not very well supported. I also don't think it creates separate containers for the different parts of Accumulo (master, tserver, gc, etc.) whereas the Gaffer one does.

Author: I'll add a comment to say its not massively supported but I do think it creates separate containers like the Gaffer one where you pass an argument to run the Accumulo service you want the container to act as see: https://github.com/apache/accumulo-docker/blob/main/README.md?plain=1#L95-L104

| `gchq/hdfs` | A custom image for running HDFS (Hadoop file system) via a container. Contains an official release of [Apache Hadoop](https://hadoop.apache.org/) which is used as the scalable data storage for Accumulo. |

> | `gchq/gaffer` | This is the main container image for Gaffer that is built on on top of the `gchq/accumulo` image so includes a release of `zookeeper`, `hdfs` and `accumulo` along with the Gaffer libraries. Running this image simply runs the Accumulo instance not a Gaffer instance. |

Reviewer 3: This runs an instance of Accumulo with the Gaffer add-ons that let you create a graph in Accumulo.
